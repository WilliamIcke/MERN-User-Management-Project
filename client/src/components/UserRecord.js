import React from 'react';
import { format } from 'date-fns';
import EditUserRecord from './EditUserRecord';
import Modal from "react-modal";

Modal.setAppElement("#root");

const UserRecord = ({ userRecord, membershipRecords, updateUserRecordItem, deleteUserRecord }) => {
    const [isOpen, setIsOpen] = React.useState(false);

    /* If a date is set, convert it to a nice format */
    let betterDateOfBirthFormat = "NA";
    if ('dateOfBirth' in userRecord && userRecord.dateOfBirth !== null) {
        betterDateOfBirthFormat = new Date(userRecord.dateOfBirth);
        betterDateOfBirthFormat = format(betterDateOfBirthFormat, 'dd/MM/yyyy');
    }

    let betterMembershipStartFormat = "NA";
    if ('membershipStart' in userRecord && userRecord.membershipStart !== null) {
        betterMembershipStartFormat = new Date(userRecord.membershipStart);
        betterMembershipStartFormat = format(betterMembershipStartFormat, 'dd/MM/yyyy');
    }

    /* Simple function to help with the Edit To Do modal */
    function toggleModal() {
        setIsOpen(!isOpen);
    }

    return (
        <div className="userRecordCard">
            <dl>
                <dt>Firstname</dt>
                <dd>{userRecord.firstname}</dd>
                <dt>Surname</dt>
                <dd>{userRecord.surname}</dd>
                <dt>Date of birth</dt>
                <dd>{betterDateOfBirthFormat}</dd>
                <dt>Membership</dt>
                <dd>{userRecord.membership.membershipName}</dd>
                <dt>Membership start date</dt>
                <dd>{betterMembershipStartFormat}</dd>
            </dl>
            <div className='Card-Actions'>
                <button onClick={toggleModal}>Edit User</button>
                <Modal
                    isOpen={isOpen}
                    onRequestClose={toggleModal}
                    contentLabel="editContent"
                >
                    <div className='editModalContent'>
                        <EditUserRecord userRecord={userRecord} membershipRecords={membershipRecords} updateUserRecordItem={updateUserRecordItem}/>
                        <button className="closeEditModal" onClick={toggleModal}>Close</button>
                    </div>
                </Modal>
            </div>
        </div>
    )
}

export default UserRecord