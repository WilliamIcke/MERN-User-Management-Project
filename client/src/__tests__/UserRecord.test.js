import React from 'react';
import { shallow, mount, render } from 'enzyme';
import UserRecord from '../components/UserRecord';

const testData = {
  userRecord: {
      _id: 1,
      firstname: 'Test',
      surname: 'Tester',
      dateOfBirth: '12/12/1990',
      membership: 1,
      membershipStart: '12/07/2018',
      active: true,
  },
  membershipRecords: [
      {
      _id: 1,
      membershipName: 'No Membership',
      cost: '0',
      active: true
      },
      {
      _id: 2,
      membershipName: 'Priority Membership',
      cost: '250',
      active: true
      },
      {
      _id: 3,
      membershipName: 'VIP Membership',
      cost: '500',
      active: true
      },
      {
      _id: 4,
      membershipName: 'Tier 1 Membership',
      cost: '120',
      active: true
      },
      {
      _id: 5,
      membershipName: 'Basic Membership',
      cost: '50',
      active: true
      }
  ]};

describe("Component shallow snapshot tests (debug)", () => {
  test('Check UserRecord component, only', () => {
    const component =  shallow(
      <UserRecord 
        key={testData.userRecord._id}
        userRecord={testData.userRecord}
        membershipRecords={testData.membershipRecords} debug/>
    );
    expect(component).toMatchSnapshot();
  });
});

describe("Functionality elements", () => {
  test('Selecting edit button triggers the modal containing edit user component', () => {
    const consoleLogSpy = jest.spyOn(console, 'log');
    const component =  shallow(
      <UserRecord 
        key={testData.userRecord._id}
        userRecord={testData.userRecord}
        membershipRecords={testData.membershipRecords} debug/>
    );
    const button = component.find('.editUserButton');
    button.simulate('click');
    expect(consoleLogSpy).toHaveBeenCalledWith('Edit user modal triggered');
  });
});

describe("Functionality elements alongside EditUserRecordModal", () => { 
   test('Selecting the close modal button should trigger the modal', () => {
    const consoleLogSpy = jest.spyOn(console, 'log');
    const component =  shallow(
      <UserRecord 
        key={testData.userRecord._id}
        userRecord={testData.userRecord}
        membershipRecords={testData.membershipRecords} debug/>
    );
    const button = component.find('.editUserButton');
    button.simulate('click');
    expect(consoleLogSpy).toHaveBeenCalledWith('Edit user modal triggered');
   });

  //  test('Edit a user record via the modal, then close the modal. No changes should be saved', () => {
        // Select the edit user button

        // Edit input fields

        // Select the close modal button

        // Check that data hasn't changed
  //  });

  //  test('Edit a user record via the modal, then save, This will close the modal and update the user record', () => {
        // Select the edit user button

        // Edit input fields

        // Select the update button

        // Check that data has changed (Including membership start data if membership changed)
  //  });
});