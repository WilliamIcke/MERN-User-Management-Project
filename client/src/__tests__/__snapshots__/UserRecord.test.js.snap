// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component shallow snapshot tests (debug) Check UserRecord component, only 1`] = `
<div
  className="userRecordCard"
>
  <dl>
    <dt>
      Firstname
    </dt>
    <dd>
      Test
    </dd>
    <dt>
      Surname
    </dt>
    <dd>
      Tester
    </dd>
    <dt>
      Date of birth
    </dt>
    <dd>
      12/12/1990
    </dd>
    <dt>
      Membership
    </dt>
    <dd />
    <dt>
      Membership start date
    </dt>
    <dd>
      07/12/2018
    </dd>
  </dl>
  <div
    className="Card-Actions"
  >
    <button
      className="editUserButton"
      onClick={[Function]}
    >
      Edit User
    </button>
    <Modal
      ariaHideApp={true}
      bodyOpenClassName="ReactModal__Body--open"
      closeTimeoutMS={0}
      contentElement={[Function]}
      contentLabel="editContent"
      isOpen={false}
      onRequestClose={[Function]}
      overlayElement={[Function]}
      parentSelector={[Function]}
      portalClassName="ReactModalPortal"
      preventScroll={false}
      role="dialog"
      shouldCloseOnEsc={true}
      shouldCloseOnOverlayClick={true}
      shouldFocusAfterRender={true}
      shouldReturnFocusAfterClose={true}
    >
      <div
        className="editModalContent"
      >
        <EditUserRecord
          membershipRecords={
            Array [
              Object {
                "_id": 1,
                "active": true,
                "cost": "0",
                "membershipName": "No Membership",
              },
              Object {
                "_id": 2,
                "active": true,
                "cost": "250",
                "membershipName": "Priority Membership",
              },
              Object {
                "_id": 3,
                "active": true,
                "cost": "500",
                "membershipName": "VIP Membership",
              },
              Object {
                "_id": 4,
                "active": true,
                "cost": "120",
                "membershipName": "Tier 1 Membership",
              },
              Object {
                "_id": 5,
                "active": true,
                "cost": "50",
                "membershipName": "Basic Membership",
              },
            ]
          }
          userRecord={
            Object {
              "_id": 1,
              "active": true,
              "dateOfBirth": "12/12/1990",
              "firstname": "Test",
              "membership": 1,
              "membershipStart": "12/07/2018",
              "surname": "Tester",
            }
          }
        />
        <button
          className="closeEditModal"
          onClick={[Function]}
        >
          Close
        </button>
      </div>
    </Modal>
  </div>
</div>
`;
